### DISCUSSION TITLE
Test Ideas for Messaging Systems & Message Queues

### DISCUSSION THREAD

**Discussion by TheENGR** [CS-6381/FinalProject#4] [Ideas]
Hey all, lets come up with some basic ideas for tests!

**Comment by jmbeach** [comment]
I'm taking a lot of "inspiration" from [this article](https://softwaremill.com/mqperf/)

# Quantitative

* [Throughput](https://en.wikipedia.org/wiki/Throughput)
    * How many messages can be processed per per second
        * How many can be sent per second
        * How many can be received per second
* Processing Latency (End-to-end)
    * Duration from the time the message is sent to the time the message is received
        * Reveals how fast the broker is
* Send Latency (Beginning-to-middle)
    * How long it takes for a message to be sent just to the queue
        * Not all the way to the consumer

## Configurations

1) 1 producer, 1 consumer
2) 1 producer, 2 consumers
3) 3 producers, 3 consumers (clustered / replicated if applicable)
4) 5 producers, 10 consumers (clustered / replicated if applicable)
5) 25 producers, 100 consumers (clustered / replicated if applicable)

Note: The larger configurations could use threading instead of hosts to achieve the number of producers/consumers

I think it makes sense for there to be more consumers than producers, but let me know if that's a bad assumption.

Edit: In each of the above scenarios, the producers and consumers are sending/receiving messages in `while true` fashion (so as much as possible). We will need to decide on:

- Message size (could be random)
- Whether or not to batch messages

**Reply by jmbeach** [reply]
I want to figure out the qualitative metrics everyone should be recording as well, but starting with this for now. These are all suggestions so let me know what you think

**Reply by thesammiller** [reply]
I think this looks pretty good. I don't know about the assumption that there will be more consumers than producers. I think that might be on a case-by-case basis. I'm thinking of IoT data devices that might send a lot of information more than receive, but I don't know. It might expose something interesting to test both sides - imbalanced producers vs. imbalanced consumers.

**Reply by rich-java-dev** [reply]
A few suggestions taking Sam's idea of imbalancing into account.

1P1C
1P2C

2P1C
2P2C

10P 5C
5P 10C

100P 5C
5P 100C

**Reply by rich-java-dev** [reply]
RTTs can be measured with a tool like PyShark and plotted with matplotlib.
https://pypi.org/project/pyshark/

Something like this to filter/analyze tcp packets, may allow testing of packet times (such as RTTs, round trip times from SYN to ACK) independently of the implementing MQ.

https://community.sinefa.com/hc/en-us/articles/211664307-TCP-Health-and-Application-Round-Trip-Time-RTT-#:~:text=RTT%2C%20or%20Round%2Dtrip%20time,every%20TCP%20connection%20Sinefa%20monitors.&text=A%20larger%20RTT%20indicates%20a%20larger%20latency.

From previous assignment, the script is here:

https://github.com/rich-java-dev/cs6381-assignment2/blob/main/src/monitor.py

What may be nice about this is a common measurement from tcp packet analysis to be ran over basic pub/sub set-ups, and time to tweak the configuration of the MQ software vs. checking the time-stamps and  compare of Pub to Sub manually in the Pubs and Subs code.

It was easy to deploy this monitor.py class in its own python runtime independently of ZMQ/assignments.

![image](https://user-images.githubusercontent.com/32989822/112733474-a974f900-8f16-11eb-9037-2a757cc7faed.png)

**Reply by jmbeach** [reply]
@rich-java-dev The one thing I worry about with wireshark is hidden delays in the tooling. The network request may only take x amount of time, but what if there's some behind the scenes processing we're missing.

**Reply by rich-java-dev** [reply]
I think thats a fair observation.

Taking that into account, it may be valuable data to collect both a time-stamp of the Pub-side client-API just before sending, and a time-stamp of the Sub-side client-API just after receiving.

Using this in conjunction with the TCP packet delays, we more or less subtract the TCP time from the client/server processing times, to possibly get better insight into the processing time the MQ spends outside of the constraints of the network transfer

![image](https://user-images.githubusercontent.com/32989822/112771600-93d8ff80-8ffa-11eb-9096-09f098fbe320.png)


Even still just analyzing T2-T1 (Pub processing) and T4-T3 (Sub processing) and combining them, without including TCP time may be useful to help isolate for better comarisons if tests are ran on slightly different hardware, or even used as a sort of Factor to normalize the results.  
EG: if we expect the actual TCP transfer time of the same data to be consistent across different MQs, then that may help us fit our data better.
It may also just help us understand which technologies have more optimized network transfers vs. entire process time.

I know this is as bit crude and doesn't take into account Broker, and you could also break down another layer  and split the code of the experiments from the standard/popular Python Client API for the MQ system, but just some food for thought.

**Reply by jmbeach** [reply]
I like it @rich-java-dev

**Comment by laynemoseley** [comment]
Two interesting examples of potential real world applications of message queues come to mind:

**YouTube**
When videos are uploaded to YouTube, there is a _lot_ of post processing that occurs, such as transcoding and copyright protection. Video transcoding especially takes a very long time as it is processor intensive. I imagine that the flow is something like this:

video entry is recorded in database -> video file is uploaded -> transcode request placed into a message queue -> transcoders pull those messages of the queue and transcode the video -> transcode complete, message queue receive `ack`, video entry in db updated

In the event the transcoding fails, the message queue does not receive an `ack` and the message stays on the queue. Another transcoder picks up the message and attempts the transcode until successful.

I don't know at all how this really works at YouTube, but it could be something like this to ensure speed and reliability.

**Stripe**
Every transaction that stripe processes hits a webhook that you provide. The webhook will hit your backend a few times, with exponential backoff, before giving up. This flow seems like it could be something like this:

payment is processed -> webhook request enters a message queue -> webhook sender pulls message off the queue, hits the webhook -> if successful, message queue received an `ack`

If the webhook fails, the message would remain on the queue until:
- the webhook is successful, or
- the maximum retry count is reached

It would be valuable to mock/simulate a flow like this and see what type of volume can be handled in different load scenarios. For example: How large of a RabbitMQ cluster is needed to process 1 million payments per day?

**Reply by jmbeach** [reply]
I like this metric. It kind of condensed all of the information like throughput into one data point

**Reply by jmbeach** [reply]
Since we also have event streams in the mix, it would probably be good to have another use case like logging

**Reply by jmbeach** [reply]
This is also something where we could still do an incremental approach.

You need x hardware/software complexity to process 10000 payments 

You need y to process 100000

**Reply by thesammiller** [reply]
I like Stripe as a metric. I'm curious about video encoding, but I don't know much about it. 

To @jmbeach point, we could check different configurations of Chameleon VMs under different loads. 10000 on Medium vs. Large, 10000 on Medium vs. Large.
